### PR TITLE
[5.5] Add support for singular resource routes.

### PR DESCRIPTION
**Pull request by patrickcarlohickman**
I targeted the master branch for 5.5, but this should be a non-breaking change and could be added to 5.4, I believe (without the changes to `PendingResourceRegistration`, which doesn't exist in 5.4).

## Overview

This PR adds the ability to generate resource routes for a singular resource.  This was discussed a long time ago in issue #1192, and received positive feedback, but nothing ever came of it.

Sometimes you need to deal with a resource that does not need to be referenced by the id. For example, you may have a `profile` resource that belongs to the currently authenticated user. Currently, if you wanted to `show` the user's profile, the route would be `/profile/{profile}`. However, since you don't need the id to look up the current user's profile, it would be nice to be able to simply have the `show` route be `/profile`. This is the same for the `edit`, `update`, and `destroy` routes, as well.

The singular resource route generates the following routes:

| Verb | Path | Action | Route Name |
| -- | -- | -- | -- |
| GET | /profile/create | create | profile.create |
| POST | /profile | store | profile.store |
| GET | /profile | show | profile.show |
| GET | /profile/edit | edit | profile.edit |
| PUT/PATCH | /profile | update | profile.update |
| DELETE | /profile | destroy | profile.destroy |

These routes differ from the current resource routes in two ways:
1. The `show`, `edit`, `update`, and `destroy` routes do not need a route parameter.
2. Since this is a singular resource, there is no `index` action.

## How It Works

In order to create a new singular resource route, this PR adds a new `singular` option to the resource route definition. This option can be set to three different values:
1. The boolean `true`. This is mainly for non-nested resource definitions. If it is used on a nested resource definition, only the last resource in the chain will be treated as singular.
2. A string containing the name of the singular resource. This can be used when there is only one resource that is singular in a nested resource definition. The string must contain the name of the singular resource.
3. An array of strings containing the names of the singular resources. This can be used in any scenario, but it must be used when there are multiple singular resources in a nested resource definition.

## Examples

**A singular resource (`/profile`):**
```
// Since there is only one singular resource, and it is the last one,
// all three options work in this scenario.

Route::resource('profile', 'ProfileController', ['only' => ['show'], 'singular' => true]);
Route::resource('profile', 'ProfileController', ['only' => ['show'], 'singular' => 'profile']);
Route::resource('profile', 'ProfileController', ['only' => ['show'], 'singular' => ['profile']]);
```

("only" removed from here on out for brevity)

**A singular resource nested under a plural resource (`/users/{user}/profile`):**
```
// Since there is only one singular resource, and it is the last one,
// all three options work in this scenario.

Route::resource('users.profile', 'ProfileController', ['singular' => true]);
Route::resource('users.profile', 'ProfileController', ['singular' => 'profile']);
Route::resource('users.profile', 'ProfileController', ['singular' => ['profile']]);
```

**A singular resource nested under a singular resource (`/profile/avatar`):**
```
// Since there are multiple resources that are singular, only the array
// syntax can be used to specify all the singular resources.

Route::resource('profile.avatar', 'ProfileAvatarController', ['singular' => ['profile', 'avatar']]);
```

**A plural resource nested under a singular resource (`/profile/phones/{phone}`):**
```
// Since there is a parent resource that is singular, the "true" option cannot be used.
// But since there is only one singular resource, the string option can be used.

Route::resource('profile.phones', 'PhoneController', ['singular' => 'profile']);
Route::resource('profile.phones', 'PhoneController', ['singular' => ['profile']]);
```

**A singular resource nested under a plural resource nested under a singular resource (`/profile/phones/{phone}/type`):**
```
// Since there are multiple resources that are singular, only the array
// syntax can be used to specify all the singular resources.

Route::resource('profile.phones.type', 'PhoneTypeController', ['singular' => ['profile', 'type']]);
```

## [5.5] Fluent Routes

Additonally, a new `singular()` method was added to the fluent route definitions. If nothing is passed in, it will default the `singular` option to `true`, otherwise it sets the `singular` option to whatever is passed in.

**A singular resource defined fluently (`/profile`):**
```
Route::resource('profile', 'ProfileController')->only('show')->singular();
Route::resource('profile', 'ProfileController')->only('show')->singular(true);
Route::resource('profile', 'ProfileController')->only('show')->singular('profile');
Route::resource('profile', 'ProfileController')->only('show')->singular(['profile']);
```

Please let me know if there are any changes that need to be made.

Thanks!

### CODE REVIEW

File: src/Illuminate/Routing/PendingResourceRegistration.php [hunk 1]
```diff
@@ -143,6 +143,19 @@ public function middleware($middleware)
     }
 
     /**
+     * Set the resource to be a singular resource instance.
+     *
+     * @param  bool|string|array  $resources
+     * @return \Illuminate\Routing\PendingResourceRegistration
```
Comment (lucasmichot) [inline src/Illuminate/Routing/PendingResourceRegistration.php]: `@return $this`

File: src/Illuminate/Routing/ResourceRegistrar.php [hunk 2]
```diff
@@ -85,12 +99,24 @@ public function register($name, $controller, array $options = [])
...
 
+        $resources = explode('.', $name);
+
+        // If the singular option is set, we need to determine which resources are meant
+        // to be singular resources. If set to true, then only the last resource will
+        // be singular, otherwise intersect all resources with the given resources.
+        if (! empty($options['singular'])) {
+            $this->singularResources = array_intersect(
+                $resources,
+                $options['singular'] === true ? [last($resources)] : array_wrap($options['singular'])
```
Comment (lucasmichot) [inline src/Illuminate/Routing/ResourceRegistrar.php]: `Arr::wrap`

File: src/Illuminate/Routing/ResourceRegistrar.php [hunk 3]
```diff
@@ -315,18 +356,22 @@ protected function getNestedResourceUri(array $segments)
...
      * Format a resource parameter for usage.
      *
      * @param  string  $value
-     * @return string
+     * @return string|null
      */
     public function getResourceWildcard($value)
     {
+        if (in_array($value, $this->singularResources)) {
+            return null;
```
Comment (lucasmichot) [inline src/Illuminate/Routing/ResourceRegistrar.php]: `return;`

File: tests/Routing/RoutingRouteTest.php [hunk 4]
```diff
@@ -1070,6 +1070,111 @@ public function testResourceRouting()
...
+
+        $router = $this->getRouter();
+        $router->resource('foo', 'FooController', ['singular' => ['foo']]);
+        $routes = $router->getRoutes();
+        $this->assertCount(6, $routes);
+
+        $router = $this->getRouter();
+        $router->resource('foo', 'FooController', ['only' => ['index'], 'singular' => true]);
+        $routes = $router->getRoutes();
+        $this->assertCount(0, $routes);
```
Comment (lucasmichot) [inline tests/Routing/RoutingRouteTest.php]: `assertEmpty`

**Comment by lucasmichot** [comment]
Does it also work with `Route::apiResource('profile',  ....)` ?
If so, it would be worth adding tests for that too

File: src/Illuminate/Routing/PendingResourceRegistration.php (hunk 1 above)
Comment (patrickcarlohickman) [inline src/Illuminate/Routing/PendingResourceRegistration.php]: All of the other `@return` statements for fluent methods in this class use the FQCN. Using `@return $this` on this method would be inconsistent with the rest of the class.

We could change all the rest of the methods to use `@return $this`, but I think that would belong in another PR.

File: src/Illuminate/Routing/ResourceRegistrar.php (hunk 2 above)
Comment (patrickcarlohickman) [inline src/Illuminate/Routing/ResourceRegistrar.php]: Thanks. Will change.

File: src/Illuminate/Routing/ResourceRegistrar.php (hunk 3 above)
Comment (patrickcarlohickman) [inline src/Illuminate/Routing/ResourceRegistrar.php]: Thanks. Will change.

File: tests/Routing/RoutingRouteTest.php (hunk 4 above)
Comment (patrickcarlohickman) [inline tests/Routing/RoutingRouteTest.php]: Thanks. Will change.

**Comment by patrickcarlohickman** [comment]
@lucasmichot Thanks for the code review!

Since `apiResource()` just calls `resource()` under the hood, this will work. I can add some tests, though.

**Comment by patrickcarlohickman** [comment]
@lucasmichot I have updated the PR with some of the code changes, as well as the additional tests for `apiResource()`.

Do you know if there is any type of coding guide available for the framework? Something that can be referred to let contributors know about some of the issues that were pointed out here (prefer `Arr`/`Str` classes over helper functions, prefer `return;` over `return null;`, etc.).

Thanks again.

**Comment by lucasmichot** [comment]
> @lucasmichot I have updated the PR with some of the code changes, as well as the additional tests for apiResource()

👍 

> Do you know if there is any type of coding guide available...

Actually not, StyleCI does most of the work but some other things cannot be checked.
I believe this is something that should be written in contributing guidelines maybe

**Comment by taylorotwell** [comment]
Holding off on this for now.

**Comment by patrickcarlohickman** [comment]
@taylorotwell I know I submitted this right before LaraconUS/EU and 5.5 release. Any chance of revisiting this? Thanks!

**Comment by idragon81** [comment]
Would love to see this

**Comment by patrickcarlohickman** [comment]
@idragon81 @taylorotwell I can submit another PR if there's interest, but in the meantime, I created this functionality as a package that I use when the PR was rejected: https://github.com/shiftonelabs/laravel-singular-resource-routes

**Comment by desertnightsky** [comment]
Need time to fulfill maintence for night and reply with response. 

On Dec 16, 2018, at 8:43 PM, Patrick Carlo-Hickman <notifications@github.com> wrote:

@idragon81 @taylorotwell I can submit another PR if there's interest, but in the meantime, I created this functionality as a package that I use when the PR was rejected: https://github.com/shiftonelabs/laravel-singular-resource-routes

**Comment by idragon81** [comment]
Using it right now, it's great. However would love to see this in the framework

**Comment by patrickcarlohickman** [comment]
@idragon81, in case you're still using Laravel and using the shiftonelabs package, it looks like this functionality was just included into Laravel 9x.

**Comment by CraftLogan** [comment]
The pr that is now merged into 9.x can be seen here https://github.com/laravel/framework/pull/44872

Thanks @patrickcarlohickman and @jessarcher for doing the work on this...
